_posts/2020-04-28-mocking.adoc

```diff
@@ -0,0 +1,375 @@
+
+    @Inject
+    MockableBean2 mockableBean2;
+
+    @BeforeAll
+    public static void setup() {
+        MockableBean1 mock = Mockito.mock(MockableBean1.class);  // <1>
+        Mockito.when(mock.greet("Stuart")).thenReturn("A mock for Stuart");
+        QuarkusMock.installMockForType(mock, MockableBean1.class);  // <2>
+    }
+
+    @Test
+    public void testBeforeAll() {
+        Assertions.assertEquals("A mock for Stuart", mockableBean1.greet("Stuart"));  // <3>
+        Assertions.assertEquals("Hello Stuart", mockableBean2.greet("Stuart")); // <4>
+    }
+
+    @Test
+    public void testPerTestMock() {
+        QuarkusMock.installMockForInstance(new BonjourMockableBean2(), mockableBean2); // <5>
+        Assertions.assertEquals("A mock for Stuart", mockableBean1.greet("Stuart"));  // <6>
+        Assertions.assertEquals("Bonjour Stuart", mockableBean2.greet("Stuart")); // <7>
+    }
+
+    @ApplicationScoped
+    public static class MockableBean1 {
+
+        public String greet(String name) {
+            return "Hello " + name;
+        }
+    }
+
+    @ApplicationScoped
+    public static class MockableBean2 {
+
+        public String greet(String name) {
+            return "Hello " + name;
+        }
+    }
+
+    public static class BonjourMockableBean2 extends MockableBean2 {
+        @Override
+        public String greet(String name) {
+            return "Bonjour " + name;
+        }
+    }
+}
+----
+<1> This part of the  example uses Mockito for convenience's sake only. `QuarkusMock` is not tied to Mockito in any way.
+<2> We use `QuarkusMock.installMockForType()` because the injected bean instance is not yet available. Very important to note is that
+the mock setup in a JUnit `@BeforeAll` method, is used for *all* test methods of the class (other test classes are *not* affected by this).
+<3> The mock for `MockableBean1` is being used as it was defined for all test methods of the class.
+<4> Since no mock has been set up for `MockableBean2`, the CDI bean is being used.
+<5> We use `QuarkusMock.installMockForInstance()` here because inside the test method, the injected bean instance is available.
+<6> The mock for `MockableBean1` is being used as it was defined for all test methods of the class.
+<7> As we used `BonjourMockableBean2` as a mock `MockableBean2`, this class is now used.
+
+[IMPORTANT]
+====
+`QuarkusMock` can be used for any normal scoped CDI bean - the most common of which are `@ApplicationScoped` and `@RequestScoped`.
+This means that beans with `@Singleton` and `@Dependent` scope *cannot* be used with `QuarkusMock`.
+
+Furthermore, `QuarkusMock` will not work properly when it's used in tests that run parallel in the same JVM.
+====
+
+Returning to the original example of the blog post, we could get rid of the `MockInvoiceNotificationService` class and
+instead use something like the following:
+
+[source,java]
+----
+public class OrderServiceTest {
+
+    @Inject
+    private OrderService orderService;
```

Comment:
We encourage users not to use `@Inject private`....

Reply:
Right! I'll fix it